### PR TITLE
Docs: Extend `PruningPredicate` with background and implementation info

### PR DESCRIPTION
**Pull request by alamb**
~Draft as it builds on https://github.com/apache/arrow-datafusion/pull/9183~

## Which issue does this PR close?

related to #9171 

## Rationale for this change

We use the `PruningPredicate` logic extensively in InfluxDB and we have found it is not an easy concept to understand. I have struggled to explain it face to face concisely so I  wrote it down with the help of @appletreeisyellow 

I would like to add this content to help both our future engineers as well as the broader community understand how this code works (perhaps so they can extend it)

I also harbor an ulterior motive that anyone who works on / with DataFusion can get an accessable master class in practical database implementation techniques by reading our docs

## What changes are included in this PR?

Add background and algorithm description to `PruningPredicate` docs

## Are these changes tested?
By doc test

## Are there any user-facing changes?
More docs. No code changes

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/pruning.rs [hunk 1]
```diff
@@ -182,6 +191,179 @@ pub trait PruningStatistics {
...
+///
+/// ## SQL Filter Semantics
+///
+/// The SQL `WHERE` clause has a boolean expression, often called a filter or
+/// predicate. The semantics of this predicate are that the query evaluates the
+/// predicate for each row in the input tables and:
+///
+/// * Rows that evaluate to `true` are returned in the query results
+/// * Rows that evaluate to `false` are not returned (“filtered out” or “pruned” or “skipped”).
+/// * Rows that evaluate to `NULL` are **NOT** returned (also “filtered out”) – *this property appears many times in the discussion below*
```
Comment (appletreeisyellow) [inline datafusion/core/src/physical_optimizer/pruning.rs:247]: > Rows that evaluate to `NULL` are **NOT** returned (also “filtered out”)

Seems like this is the opposite of what discussed below. 

For example, the quote below says -- "containers that evaluate to `NULL` are returned (also 'kept')"

> /// * `NULL`: there MAY be rows that pass the predicate, **KEEPS** the container
///           Note that rewritten predicate can evaluate to NULL when some of
///           the min/max values are not known.

Did I misunderstand something here? 🤔

Comment (alamb) [reply datafusion/core/src/physical_optimizer/pruning.rs]: This is a very good observation.

Yes the NULL semantics in filter expressions is the opposite of what the rewritten pruning predicate does with NULL (which is quite confusing). I added a note trying to clarify this: 74261e8e3

File: datafusion/core/src/physical_optimizer/pruning.rs [hunk 2]
```diff
@@ -149,11 +149,12 @@ pub trait PruningStatistics {
 /// for any row in the Row Group, the entire Row Group is skipped during query
 /// execution.
 ///
-/// The `PruningPredicate` API is designed to be general, so it can used for
-/// pruning other types of containers (e.g. files) based on statistics that may
-/// be known from external catalogs (e.g. Delta Lake) or other sources.
+/// The `PruningPredicate` API is general, and can be used for pruning other
+/// types of containers (e.g. files) based on statistics that may be known from
+/// external catalogs (e.g. Delta Lake) or other sources. How his works is a
```
Comment (Jefffrey) [inline datafusion/core/src/physical_optimizer/pruning.rs:154]:
```suggestion
/// external catalogs (e.g. Delta Lake) or other sources. How this works is a
```

**Review by Jefffrey — APPROVED** [review]
Makes sense to me, thanks for the great explanation. I'm not familiar with the PruningPredicate code much, but from the documentation alone here it looks good (reviewing from phone so it's a little hard to check out the code 😅)

**Pull request merged**